File: 746-min-cost-climbing-stairs/746-min-cost-climbing-stairs.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        int zeroIndex = minCost(0, n, cost, new HashMap<Integer, Integer>());
        int firstIndex = minCost(1, n, cost, new HashMap<Integer, Integer>());
        
        return Math.min(zeroIndex, firstIndex);
    }
    private int minCost(int curIndex, int n, int[] cost, 
                        HashMap<Integer, Integer> memo) {
        //base case
        if(curIndex==n) {
            return 0;
        }
        if(curIndex>n) {
            return 1001;
        }
        
        int curKey = curIndex;
        if(memo.containsKey(curKey)) {
            return memo.get(curKey);
        }
        
        int oneJump = cost[curKey] + minCost(curKey+1, n, cost, memo);
        int twoJumps = cost[curKey] + minCost(curKey+2, n, cost, memo);
        
        memo.put(curKey, Math.min(oneJump, twoJumps));
        
        return memo.get(curKey);
    } 
}

File: 746-min-cost-climbing-stairs/746-min-cost-climbing-stairs.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        unordered_map<int, int> mp1;
        unordered_map<int, int> mp2;
        
        int ans1 = minCost(0, cost, mp1);
        int ans2 = minCost(1, cost, mp2);
        
        return min(ans1, ans2);
    }
    
    int minCost(int idx, vector<int> &cost, unordered_map<int, int> &mp)
    {
        if(idx >= cost.size()) return 0;
        
        int curKey = idx;
        if(mp.find(idx)!=mp.end())
            return mp[idx];
        
        int oneJump = cost[idx] + minCost(idx+1, cost, mp);
        int twoJumps = cost[idx] + minCost(idx+2, cost, mp);
        
        mp[curKey] = min(oneJump, twoJumps);
        return mp[curKey];
        
    }
};